### PR TITLE
0.0.0

### PR DESCRIPTION
**Pull request by splatteredbits**

### CODE REVIEW

File: Zip/Functions/New-ZipArchive.ps1 [hunk 1]
```diff
@@ -0,0 +1,85 @@
+
+function New-ZipArchive
+{
+    <#
+    .SYNOPSIS
+    Creates a new, empty ZIP archive.
+
+    .DESCRIPTION
+    The `New-ZipArchive` function createa a new, empty ZIP archive. Pass the path to the archive to the Path parameter. A new, empty ZIP archive is created at that path. The function returns a `IO.FileInfo` objecy representing the new ZIP archive.
```
Comment (decoyjoe) [inline Zip/Functions/New-ZipArchive.ps1]: Typo: `objecy` should be `object`

File: Zip/Functions/New-ZipArchive.ps1 [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+function New-ZipArchive
+{
+    <#
+    .SYNOPSIS
+    Creates a new, empty ZIP archive.
+
+    .DESCRIPTION
+    The `New-ZipArchive` function createa a new, empty ZIP archive. Pass the path to the archive to the Path parameter. A new, empty ZIP archive is created at that path. The function returns a `IO.FileInfo` objecy representing the new ZIP archive.
+
+    If `Path` is relative, it is create relative to the current directory.
```
Comment (decoyjoe) [inline Zip/Functions/New-ZipArchive.ps1]: Typo: `it is create relative`

File: Zip/Functions/New-ZipArchive.ps1 [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    
+    By default, entry names are encoded as UTF8 text. If your ZIP archive will be consumed by tools that don't support UTF8, pass the encoding they do support to the `EntryNameEncoding` parameter.
+
+    .EXAMPLE
+    New-ZipArchive -Path 'archive.zip'
+
+    Creates a new, empty ZIP file named `archive.zip` in the current directory.
+
+    .EXAMPLE
+    New-ZipArchive Path 'archive.zip' -Force
```
Comment (decoyjoe) [inline Zip/Functions/New-ZipArchive.ps1]: Typo: `Path` should be `-Path`

File: Zip/Functions/New-ZipArchive.ps1 [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    Creates a new, empty ZIP file named `archive.zip` in the current directory. If a file named 'archive.zip' already exists in the current directory, it is deleted and a new file created.
+
+    .EXAMPLE
+    New-ZipArchive Path 'archive.zip' -CompressionLevel Fastest -Encoding [Text.Encoding]::ASCII
```
Comment (decoyjoe) [inline Zip/Functions/New-ZipArchive.ps1]: Typo: `Path` should be `-Path`

File: Zip/Functions/New-ZipArchive.ps1 [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
+
+    if( -not [IO.Path]::IsPathRooted($Path) )
+    {
+        $Path = Join-Path -Path (Get-Location) -ChildPath $Path
+    }
+    $Path = [IO.Path]::GetFullPath($Path)
+
+    if( (Test-Path -Path $Path) )
```
Comment (decoyjoe) [inline Zip/Functions/New-ZipArchive.ps1:61]: This check should include `-PathType Leaf` because otherwise if the given path **exists**, is a **directory**, and **Force** was specified, we might inadvertently delete a whole directory.

File: Zip/Functions/New-ZipArchive.ps1 [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            Remove-Item -Path $Path
+        }
+        else
+        {
+            Write-Error -Message ('The file "{0}" already exists. Unable to create a new ZIP archive at that path. Use the -Force switch to overwrite the file.' -f $Path)
+            return
+        }
+    }
+
+    $tempDir = Join-Path -Path $env:TEMP -ChildPath ('{0}.{1}' -f ($Path | Split-Path -Leaf),([IO.Path]::GetRandomFileName()))
```
Comment (decoyjoe) [inline Zip/Functions/New-ZipArchive.ps1]: `$env:TEMP` is a Windows specific variable that will cause this function to fail when running on MacOS or a Linux system. We should use `[System.IO.Path]::GetTempPath()` which is [indeed cross platform](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.gettemppath?view=netcore-2.2).

File: Tests/New-ZipArchive.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+Describe 'New-ZipArchive.when file exists' {
+    Init
+    GivenFile 'somefile.zip'
+    WhenCreatingArchive 'somefile.zip' -ErrorAction SilentlyContinue
+    ThenNothingReturned
+    ThenError -Matches 'already exists'
+}
+
+Describe 'New-ZipArchive.when file exists and forcing creation' {
```
Comment (decoyjoe) [inline Tests/New-ZipArchive.Tests.ps1:127]: We should add a test for the scenario where a given path exists but it's a **directory** rather than a file. We should not allow `-Force` to delete an entire directory.

File: Zip/Functions/Add-ZipArchiveEntry.ps1 [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        [Alias('FullName')]
+        [Alias('Path')]
+        [string]
+        # The files/directories to add to the archive. Normally, you would pipe file/directory objects to `Add-ZipArchiveEntry`. You may also pass any object that has a `FullName` or `Path property. You may also pass the path as a string.
+        #
+        # If you pass a directory object or path to a directory, all files in that directory and all its sub-directories will be added to the archive.
+        $InputObject,
+
+        [string]
+        # By default, items are added to the ZIP archive using their full paths (minus any drives/qualifiers). The `BasePath` parameter controls what portion of the source file's path is removed when determining its name in the ZIP archive.
```
Comment (decoyjoe) [inline Zip/Functions/Add-ZipArchiveEntry.ps1]: This strikes me as odd that this is the default behavior. As a user, I've been accustomed to relative paths always being the default behavior for commands. I feel this should be the opposite, if a user specifies `C:\projects\zip` then the path in the ZIP archive should be simply `zip`, not `projects\zip`. This is how archive tools such as 7-zip and the Zip utility built into Windows work by default.

What if we add a `PathMode` parameter, similar to how 7-zip employs it? For example:

Given path: `C:\projects\zip`

Path Mode: `Relative`
```
zip
```
Path Mode: `Full`
```
projects\zip
```
Path Mode: `Absolute`
```
C_\projects\zip
```

Additionally with this being the default behavior, I can't collect files at different paths and add them the root of the zip:
```
'C:\dir\root.txt', 'C:\dir\subdir\sub.txt' | Add-ZipArchiveEntry -ZipArchivePath 'archive.zip' -BasePath 'C:\dir'
```
Results in:
```
root.txt
subdir\sub.txt
```
Instead of:
```
root.txt
sub.txt
```

File: Tests/Add-ZipArchiveEntry.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+        }
+
+        if( -not (Test-Path -Path $fullPath -PathType Leaf) )
+        {
+            New-Item -Path $fullPath -ItemType 'File'
+        }
+
+        if( $Content )
+        {
+            [IO.File]::WriteAllText($fullPath,$Content)
```
Comment (decoyjoe) [inline Tests/Add-ZipArchiveEntry.Tests.ps1:36]: Could you let me know what the reason/edge-case is for using `[IO.File]::WriteAllText` over `Set-Content`?

File: Tests/Add-ZipArchiveEntry.Tests.ps1 [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    ThenArchiveContains 'one.aspx','one.js'
+    ThenArchiveNotContains 'one.cs','one.txt'
+}
+
+Describe 'Add-ZipArchiveEntry.when file already exists' {
+    Init
+    GivenFile 'one.cs' 'first'
+    WhenAddingFiles '*.cs'
+    GivenFile 'one.cs' 'second'
+    WhenAddingFiles '*.cs' -ErrorAction SilentlyContinue
```
Comment (decoyjoe) [inline Tests/Add-ZipArchiveEntry.Tests.ps1:206]: No assertion, `ThenError`, for the error message this test should produce.

File: Tests/Add-ZipArchiveEntry.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+            Pop-Location
+        }
+    }
+    else
+    {
+        Find-GlobFile -Path $TestDrive.FullName -Include $Path | Add-ZipArchiveEntry @params
+    }
+}
+
+Describe 'Add-ZipArchiveEntry' {
```
Comment (decoyjoe) [inline Tests/Add-ZipArchiveEntry.Tests.ps1:193]: Additional test: What happens when `Add-ZipArchiveEntry` is passed a path that **does not** exist under the `BasePath` directory?
```
'C:\dir\file.txt', 'C:\some\other\dir\second.txt' | Add-ZipArchiveEntry -ZipArchivePath 'archive.zip' -BasePath 'C:\dir'
```

**Review by decoyjoe — CHANGES_REQUESTED** [review]

File: Zip/Functions/Add-ZipArchiveEntry.ps1 (hunk 8 above)
Comment (splatteredbits) [reply Zip/Functions/Add-ZipArchiveEntry.ps1]: I could go with the default behavior being adding relatively. If you pass in a file, then it gets added to the root of the package. If you pass in a directory, it also gets added to the root.

I don't think we need a PathMode. You can currently use `ParentEntryPath` to control an item's parent path in the archive. Given path `C:\projects\zip`, by default it would get added as zip, but the user could pass `C_\projects` or `projects` to get the `Full` and `Absolute` behavior.

Comment (decoyjoe) [reply Zip/Functions/Add-ZipArchiveEntry.ps1]: Yeah that's a simple enough solution.

A `PathMode` parameter would be additional functionality of making `EntryParentPath` be inferred automatically when used with a `Full` or `Absolute` path mode. Currently, if two items passed to the function have differing parent paths, you would need to call `Add-ZipArchiveEntry` twice with two different values for `EntryParentPath`.

For example:
```
'C:\projects\zip' | Add-ZipArchiveEntry -EntryParentPath 'projects'
'C:\other\misc\configs' | Add-ZipArchiveEntry -EntryParentPath 'other\misc'
```
The same folder structure in the archive could be inferred dynamically:
```
'C:\projects\zip', 'C:\other\misc\configs' | Add-ZipArchiveEntry -PathMode 'Full'
```

File: Tests/Add-ZipArchiveEntry.Tests.ps1 (hunk 9 above)
Comment (splatteredbits) [reply Tests/Add-ZipArchiveEntry.Tests.ps1:36]: `Get-Content`/`Set-Content` add newlines.

**Comment by splatteredbits** [comment]
Ready for another review.

File: Zip/Functions/Add-ZipArchiveEntry.ps1 [hunk 12]
```diff
@@ -61,12 +60,20 @@ function Add-ZipArchiveEntry
...
+        [Parameter(ParameterSetName='ItemName')]
+        [ValidatePattern('^[^\\/]')]
+        [ValidatePattern('[^\\/]$')]
+        [string]
+        # By default, items are added to the ZIP archive using their name. You can change the name with this parameter. For example, if you added file `Zip.psd1` and passed `NewZip.psd1` as the value to the parameter, the file would get added as `NewZip.psd1`.
+        $EntryName,
+
+        [ValidatePattern('^[^\\/]')]
         [string]
-        # A parent path to add to each file in the ZIP archive. If you pass 'package' to this parameter, and you're adding an item at 'file.txt', the file will be added to the archive as `package\file.txt`.
```
Comment (decoyjoe) [inline Zip/Functions/Add-ZipArchiveEntry.ps1]: Did you mean to remove this parameter doc entirely?

File: Zip/Functions/Add-ZipArchiveEntry.ps1 [hunk 13]
```diff
@@ -159,6 +153,53 @@ function Add-ZipArchiveEntry
...
         }
+
+        $filePaths = $InputObject | Resolve-Path | Select-Object -ExpandProperty 'ProviderPath'
+        foreach( $filePath in $filePaths )
+        {
+            Write-Debug -Message $filePath
+
+            if( $BasePath )
+            {
+                $baseEntryName = $filePath -replace $basePathRegex,''
```
Comment (decoyjoe) [inline Zip/Functions/Add-ZipArchiveEntry.ps1:170]: Something I just noticed, when we construct the `$basePathRegex` on [line 101](https://github.com/webmd-health-services/Zip/blob/8ab06e4b060144c9c083a9c4d471c742518bcab2/Zip/Functions/Add-ZipArchiveEntry.ps1#L101), we append a `[IO.Path]::DirectorySeparatorChar`. Do we need to first trim the given value, e.g. `$BasePath.Trim([IO.Path]::DirectorySeparatorChar)`? What if someone gave `C:\projects\zip\` as the `BasePath`, would the pattern become `^C:\\projects\\zip\\\\?` (extra escaped `\` char)? Or is this handled somewhere already?

File: Tests/Add-ZipArchiveEntry.Tests.ps1 [hunk 14]
```diff
@@ -245,29 +229,59 @@ Describe 'Add-ZipArchiveEntry.when adding archive root' {
...
+    GivenFile 'dir1\one.cs','dir1\two.cs', 'dir1\three\four.cs'
+    WhenAddingFiles 'dir1' -WithName '1dir'
+    ThenArchiveContains '1dir\one.cs','1dir\two.cs','1dir\three\four.cs'
+    ThenArchiveNotContains 'dir1\one.cs','dir1\two.cs','dir1\three\four.cs'
+}
+
+Describe 'Add-ZipArchiveEntry.when passing a directory with a custom base path' {
+    Init
+    GivenFile 'dir1\one.cs','dir1\two.cs', 'dir1\three\four.cs'
+    WhenAddingFiles 'dir1'
```
Comment (decoyjoe) [inline Tests/Add-ZipArchiveEntry.Tests.ps1]: Describe says `when passing a directory with a custom base path` but no base path parameter is given.

File: Tests/Add-ZipArchiveEntry.Tests.ps1 [hunk 15]
```diff
@@ -245,29 +229,59 @@ Describe 'Add-ZipArchiveEntry.when adding archive root' {
...
+    GivenFile 'dir1\another\one.cs','dir1\another\two.cs'
+    $root = Join-Path -Path $TestDrive.FullName -ChildPath 'dir1'
+    WhenAddingFiles 'dir1\another\one.cs','dir1\another\two.cs' -AtArchiveRoot 'dir2' -WithBasePath $root
+    ThenArchiveContains 'dir2\another\one.cs','dir2\another\two.cs'
+}
+
+Describe 'Add-ZipArchiveEntry.when giving a direcotry a new root name' {
     Init
     GivenFile 'dir1\one.cs','dir1\two.cs'
     $root = Join-Path -Path $TestDrive.FullName -ChildPath 'dir1'
```
Comment (decoyjoe) [inline Tests/Add-ZipArchiveEntry.Tests.ps1]: `$root` is defined but unused in this test.

**Review by decoyjoe — CHANGES_REQUESTED** [review]

File: Zip/Functions/Add-ZipArchiveEntry.ps1 (hunk 12 above)
Comment (splatteredbits) [reply Zip/Functions/Add-ZipArchiveEntry.ps1]: Oops.

File: Zip/Functions/Add-ZipArchiveEntry.ps1 (hunk 13 above)
Comment (splatteredbits) [reply Zip/Functions/Add-ZipArchiveEntry.ps1:170]: Fixed.

File: Tests/Add-ZipArchiveEntry.Tests.ps1 (hunk 15 above)
Comment (splatteredbits) [reply Tests/Add-ZipArchiveEntry.Tests.ps1]: Fixed.

**Comment by splatteredbits** [comment]
Ready for another review.

**Review by decoyjoe — APPROVED** [review]
